internal/hcsoci/assigned_devices.go

```diff
@@ -0,0 +1,251 @@
+	}
+	drivers := strings.Split(csDrivers, ",")
+	for _, driver := range drivers {
+		if _, err := os.Stat(driver); err != nil {
+			return nil, errors.Wrapf(err, "failed to find path to drivers at %s", driver)
+		}
+	}
+	return drivers, nil
+}
+
+// setupDeviceUtilTools finds the utility tools' host path, mounts it using vsmb, and
+// returns the UVM path to the tools
+func setupDeviceUtilTools(ctx context.Context, coi *createOptionsInternal, r *Resources) (string, error) {
+	toolsHostPath, err := getAssignedDeviceUtilityTools(coi)
+	if err != nil {
+		return "", err
+	}
+	resultUVMPaths, err := mountDirectoriesUVM(ctx, coi.HostingSystem, r, toolsHostPath)
+	if err != nil {
+		return "", nil
+	}
+	return resultUVMPaths[0], nil
+}
+
+// setupDriversWindows finds specified kernel driver directories, mounts them using vsmb,
+// then installs them in the UVM
+func setupDriversWindows(ctx context.Context, coi *createOptionsInternal, resources *Resources) error {
+	drivers, err := getAssignedDeviceKernelDrivers(coi)
+	if err != nil {
+		return err
+	}
+	if drivers == nil {
+		// no drivers were specified, skip installing drivers
+		return nil
+	}
+
+	driverUVMPaths, err := mountDirectoriesUVM(ctx, coi.HostingSystem, resources, drivers...)
+	if err != nil {
+		return err
+	}
+
+	if err := execPnPInstallAllDrivers(ctx, coi.HostingSystem, driverUVMPaths); err != nil {
+		return err
+	}
+	return nil
+}
+
+// addDeviceDriversToUVM mounts all specified driver files using VSMB and returns their path
+// in the UVM
+func mountDirectoriesUVM(ctx context.Context, vm *uvm.UtilityVM, r *Resources, hostPaths ...string) (resultUVMPaths []string, err error) {
+	for _, d := range hostPaths {
+		options := &hcsschema.VirtualSmbShareOptions{
+			ReadOnly:            true,
+			CacheIo:             true,
+			ShareRead:           true,
+			ForceLevelIIOplocks: true,
+			NoDirectmap:         vm.DevicesPhysicallyBacked(),
+		}
+		share, err := vm.AddVSMB(ctx, d, "", options)
+		if err != nil {
+			return nil, fmt.Errorf("failed to add VSMB share to utility VM for path %+v: %s", d, err)
+		}
+		r.resources = append(r.resources, share)
+		uvmPath, err := vm.GetVSMBUvmPath(ctx, d)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get VSMB share path in utility VM for path %+v: %s", d, err)
+		}
+		resultUVMPaths = append(resultUVMPaths, uvmPath)
+	}
+	return resultUVMPaths, nil
+}
+
+// getBusAssignedChildrenDeviceLocationPaths queries the UVM with the device-util tool with the formatted
+// parent bus device for the children devices' location paths from the uvm's view
+// Returns a slice of WindowsDevices created from the resulting children location paths
+func getBusAssignedChildrenDeviceLocationPaths(ctx context.Context, coi *createOptionsInternal, vmBusInstanceIDs []string, deviceUtilPath string) ([]specs.WindowsDevice, error) {
+	p, l, err := createNamedPipeListener()
+	if err != nil {
+		return nil, err
+	}
+	var pipeResults []string
+	done := make(chan bool)
```

Comment:
EXTREME nit: You can use an empty struct channel as its size is 0 and all we're doing is waiting to receive any value whatsoever.